docs/sphinx/_ext/autodoc_configurable.py

```diff
@@ -0,0 +1,125 @@
+import json
+import textwrap
+from typing import Any, List
+
+from dagster import BoolSource, IntSource, StringSource
+from dagster.config.config_type import ConfigType, ConfigTypeKind
+from dagster.core.definitions.configurable import ConfigurableDefinition
+from sphinx.ext.autodoc import DataDocumenter
+
+
+def type_repr(config_type: ConfigType) -> str:
+    """Generate a human-readable name for a given dagster ConfigType"""
+
+    # Use given name if possible
+    if config_type.given_name:
+        return config_type.given_name
+
+    # The xSource types are a particular kind of selector that is very common, special case these
+    if config_type == StringSource:
+        return "dagster.StringSource"
+    elif config_type == BoolSource:
+        return "dagster.BoolSource"
+    elif config_type == IntSource:
+        return "dagster.IntSource"
+    elif config_type.kind == ConfigTypeKind.ANY:
+        return "Any"
+    elif config_type.kind == ConfigTypeKind.SCALAR:
+        return config_type.scalar_kind.name.lower()
+    elif config_type.kind == ConfigTypeKind.ENUM:
+        return "Enum{" + ", ".join(str(val) for val in config_type.config_values) + "}"
+    elif config_type.kind == ConfigTypeKind.ARRAY:
+        return "List[{}]".format(type_repr(config_type.inner_type))
+    elif config_type.kind == ConfigTypeKind.SELECTOR:
+        return "selector"
+    elif config_type.kind == ConfigTypeKind.STRICT_SHAPE:
+        return "strict dict"
+    elif config_type.kind == ConfigTypeKind.PERMISSIVE_SHAPE:
+        return "permissive dict"
+    elif config_type.kind == ConfigTypeKind.SCALAR_UNION:
+        return (
+            f"Union[{type_repr(config_type.scalar_type)}, {type_repr(config_type.non_scalar_type)}]"
+        )
+    elif config_type.kind == ConfigTypeKind.NONEABLE:
+        return f"Union[{type_repr(config_type.inner_type)}, None]"
+
+
+def config_field_to_lines(field, name=None) -> List[str]:
+    """Given a config field, turn it into a list of lines to add to the documentation"""
+    lines = [""]
+
+    # The only unnamed field will be the top level config schema wrapper
+    if name:
+        type_str = type_repr(field.config_type)
```

Comment:
i think this whole implementation is already good enough so we don't have to do further polish atm, so this is a nit: 
i wonder if there's an easy to make these type <img src="https://user-images.githubusercontent.com/4531914/142975927-ba5ad8b7-6909-479f-8eb7-e366c9ac1611.png" height=100> clickable like the other types generated from autodata: <img src="https://user-images.githubusercontent.com/4531914/142975966-74636286-1412-496b-9569-f555d2000d9d.png" height=100>

Reply:
Yeah I really wanted it to work like that, but because of the RST format, you can't really put links inside of the headings like that, because the heading looks like: 

     :some_name (some_type):

 and the format for making something link to a type is:

    :py:class:`dagster.some_type`

, so if you try to put that link inside the heading the colons interact badly with each other:

    :some_name (:py:class:`dagster.some_type`):


The typical solution is to put the type and the param name in separate boxes, but I thought that was a lot more cluttered/ugly.

Crazy that there's (apparently) no reasonable fix for this.